token/src/token.rs

```diff
@@ -0,0 +1,190 @@
+use bulletproofs::{BulletproofGens, PedersenGens};
+use curve25519_dalek::ristretto::CompressedRistretto;
+use curve25519_dalek::scalar::Scalar;
+use zkvm::{
+    Commitment, Data, Entry, Input, Instruction, Predicate, Program, Prover, Signature, Tx,
+    TxHeader, TxID, TxLog, VMError, Value, VerificationKey, Verifier,
+};
+
+/// Represents a ZkVM Token with unique flavor and embedded
+/// metadata protected by a user-supplied Predicate.
+pub struct Token {
+    issuance_predicate: Predicate,
+    metadata: Vec<u8>,
+}
+
+impl Token {
+    /// Constructs a new Token.
+    pub fn new(pred: Predicate, metadata: Vec<u8>) -> Self {
+        Token {
+            issuance_predicate: pred,
+            metadata: metadata,
+        }
+    }
+
+    /// Returns program that issues specified quantity of Token.
+    pub fn issue<'a>(program: &'a mut Program, token: &Token, qty: u64) -> &'a mut Program {
+        program
+            .push(Commitment::blinded(qty)) // stack: qty
+            .var() // stack: qty-var
+            .push(Commitment::unblinded(Value::issue_flavor(
+                &token.issuance_predicate,
+                Data::Opaque(token.metadata.clone()),
+            ))) // stack: qty-var, flv
+            .var() // stack: qty-var, flv-var
+            .push(Data::Opaque(token.metadata.to_vec())) // stack: qty-var, flv-var, data
+            .push(token.issuance_predicate.clone()) // stack: qty-var, flv-var, data, flv-pred
+            .issue() // stack: issue-contract
+            .sign_tx() // stack: issued-value
+    }
+
+    /// Returns program that issues specified quantity of Token,
+    /// outputting it to the destination Predicate.
+    pub fn issue_to<'a>(
+        program: &'a mut Program,
+        token: &Token,
+        qty: u64,
+        dest: &Predicate,
+    ) -> &'a mut Program {
+        Token::issue(program, token, qty)
+            .push(dest.clone())
+            .output(1)
+    }
+
+    /// Returns program that retires a Token.
+    /// TBD: better symmetry here to accept a Token object
+    pub fn retire<'a>(
+        program: &'a mut Program,
+        qty: CompressedRistretto,
```

Comment:
maybe instead of asking for raw components of an `Input` and building it inside `retire`, ask for the `Input` itself? Or even better: ask for an iterator of `Input` (`<I: IntoIterator<Item=Input>>`) and add all the inputs and their retirement? And call the method `retire`.

Reply:
the thought behind getting the raw components of `Input` is that it's easier to get from the txlog, whereas otherwise the user has to construct the `Input` themselves — but that might be better?

Reply:
my thinking was that user would already have a store of utxos, so they'd be constructing or reading Inputs from there - not from the txlog. Txlog would be used only to confirm and record those utxos at an earlier time - when the incoming payment happens and way before the outgoing tx is formed.